### PR TITLE
Add the DELETE /v3/service_instances endpoint

### PR DESCRIPTION
**Pull request by mnitchev**
## Is there a related GitHub Issue?
#617 

## What is this change about?
Implements delete for service instances.

## Does this PR introduce a breaking change?
No

## Acceptance Steps
See issue #617 

## Tag your pair, your PM, and/or team
@cloudfoundry/eirini

### CODE REVIEW

File: api/repositories/service_instance_repository.go
Comment (tcdowney) [inline api/repositories/service_instance_repository.go:166]: This is an example of what we discussed a few weeks back right? Since we don't know what K8s namespace the instance lives in we have to list so we're optimizing by doing that with the privileged client since it is able to maintain a cache.

I think that's fine and good, but I think the way we're doing it here is risky since we're actually using the return value of the privileged client call and only using the userClient for authz.

I think it would be safer if we refactored the privileged client actions into a function (like `privilegedLookupNamespaceForServiceInstance` (I'm sure y'all can come up with a better name 😅 )) that only returns the namespace and then we rely on the userClient for actually fetching the returned object.

File: api/repositories/service_instance_repository_test.go
Comment (tcdowney) [inline api/repositories/service_instance_repository_test.go:448]: Nit: Even though there is no additional set up needed, I think it would improve clarity if we wrapped this in a `When("the user does not have permissions to get the service instance", ...`

File: api/repositories/service_instance_repository.go
Comment (tcdowney) [reply api/repositories/service_instance_repository.go:166]: I know there's test that verify the insufficient permissions case, but I think this will both provide additional context for what is going on (via the function name) and give us a little more safety in the future.

**Review by tcdowney — CHANGES_REQUESTED** [review]
Overall looks good. I would like to see some changes to the privileged vs user client calls in the repo though. Thanks!

File: api/repositories/service_instance_repository_test.go
Comment (danail-branekov) [reply api/repositories/service_instance_repository_test.go:448]: I would strongly vote against such contexts as they do not contribute to making the test environment more specific, instead they are (ab)used for documentation purposes. If documenting user not being allowed to do that is really desired, I would prefer putting that into the `It` description, e.g. `It("returns a forbidden error (as the user is not allowed to do that)....`

File: api/apis/service_instance_handler.go
Comment (danail-branekov) [inline api/apis/service_instance_handler.go:208]: Generally, the `Get-Delete` pattern when dealing with K8S is racy. I wonder whether we can implement the handler just call the repository's `Delete` method and handle an eventual not found error appropriately. The only thing the service instance found is used for is to get the space GUID, but on the other hand, the service instance GUID should be unique and therefore sufficient to locate and delete it.

File: api/repositories/service_instance_repository.go
Comment (danail-branekov) [reply api/repositories/service_instance_repository.go:166]: +1 for returning a record for the object being found by the unprivileged client

File: api/tests/e2e/service_instances_test.go
Comment (danail-branekov) [inline api/tests/e2e/service_instances_test.go:13]: As this PR introduces a `get service` endpoint, there should be an e2e test for it as well (provided that we need `Get` at all, see my previous comments)

File: api/repositories/service_instance_repository_test.go
Comment (tcdowney) [reply api/repositories/service_instance_repository_test.go:448]: 🤷 

Maybe it's due to my background in the Ruby community, but the most common style guides recommend this. I read  code better when I can scan up and down through the contexts to see what is being tested, but I see how this can be a personal preference.

What I'm suggesting is what is recommended by these guidelines and I generally agree:
* https://www.betterspecs.org/#contexts
* https://www.betterspecs.org/#short

That said, I also don't like to see deeply nested contexts for no reason, so it's definitely a judgement call. I don't think what I suggested _needs_ to be done in this case, but I also wouldn't want to see a hard rule against what I'm suggesting because I personally find it better. 🙂

Comment (georgethebeatle) [reply api/repositories/service_instance_repository_test.go:448]: The closes example that I can think of is [this](https://medium.com/swlh/effective-ginkgo-gomega-b6c28d476a09#cb3d) by one of Ginkgo's maintainers. I would suggest that you put a comment in the ginkgo style guide proposal, since it sounds related

Comment (tcdowney) [reply api/repositories/service_instance_repository_test.go:448]: Yeah I added a comment on the proposal yesterday 👍 

My interpretation of that blog post is that it's centered around the role of `BeforeEach` and less around using `Describe` and `Context` to provide test clarity.

Overall my thinking here though is that it's not a huge deal either way and I don't think there is much value in policing this issue across the whole codebase. I think either way has the potential to result in readable and functioning code so it's ok for something like this to differ based on who wrote the test.

I think on the proposal I agreed with the particular example that something like `When("on the happy path", ...` is not helpful and should be avoided, but that if someone adds a test and has something like `When("the user is unauthorized", ...` and does not need to provide additional setup because it's a lack of setup then that should be perfectly fine to do so. On the flipside I see that others disagree so I retract my suggestion here of asking for it to be done.

File: api/tests/e2e/service_instances_test.go
Comment (georgethebeatle) [reply api/tests/e2e/service_instances_test.go:13]: We are only introducing the ability to get a service instance on the repo level. There is no endpoint for getting the service instance on the handler level, so it is not possible to add an e2e test for this.

Comment (tcdowney) [reply api/tests/e2e/service_instances_test.go:13]: I don't think we'll be adding that endpoint soon (`GET /v3/service_instances/:guid`) since the CLI is always fetching these via list calls (filtered by `name` and `space_guid`) -- at least for the CLI commands we're targeting right now.

**Review by tcdowney — APPROVED** [review]
LGTM once conflicts are resolved. The namespaceGetting implementation might need to eventually be reconciled with the changes suggested in [this draft PR](https://github.com/cloudfoundry/cf-k8s-controllers/pull/689), but that's still a draft and gathering opinions.

File: api/apis/service_instance_handler.go
Comment (georgethebeatle) [reply api/apis/service_instance_handler.go:208]: That is a very good point @tcdowney I will address it today. This is also a very good example of why we need to handle our errors in a more generic way to avoid overlooking corner cases like this one in future.

**Review by tcdowney — APPROVED** [review]
LGTM w/ latest changes and conflicts fixed.

**Pull request merged**
